File: src/model/nostr/network/subscriptions/subscription.go
//go:build js && wasm
// +build js,wasm

package subscriptions

import (
	"context"
	"sync"
	"time"

	"github.com/candypoets/nutscash/nostr/types"
)

// subscription implements the Subscription interface
type subscription struct {
	id         string
	ctx        context.Context
	cancelFunc context.CancelFunc
	sentEvents map[string]*[]types.ParsedEvent
	mutex      sync.RWMutex
	createdAt  time.Time
	
	// FETCHED_EVENT batching fields
	fetchedBatch    [][]types.ParsedEvent
	batchingMode    bool
}

// NewSubscription creates a new subscription
func NewSubscription(id string) Subscription {
	ctx, cancel := context.WithCancel(context.Background())
	return &subscription{
		id:              id,
		ctx:             ctx,
		cancelFunc:      cancel,
		sentEvents:      make(map[string]*[]types.ParsedEvent),
		createdAt:       time.Now(),
		fetchedBatch:    make([][]types.ParsedEvent, 0),
		batchingMode:    true, // Start in batching mode
	}
}

func (s *subscription) ID() string {
	return s.id
}

func (s *subscription) Cancel() {
	s.cancelFunc()
}

func (s *subscription) IsCancelled() bool {
	select {
	case <-s.ctx.Done():
		return true
	default:
		return false
	}
}

func (s *subscription) GetSentEvents() map[string]*[]types.ParsedEvent {
	s.mutex.RLock()
	defer s.mutex.RUnlock()

	// Return a copy to prevent external mutation
	result := make(map[string]*[]types.ParsedEvent, len(s.sentEvents))
	for k, v := range s.sentEvents {
		if v != nil {
			eventsCopy := make([]types.ParsedEvent, len(*v))
			copy(eventsCopy, *v)
			result[k] = &eventsCopy
		}
	}
	return result
}

func (s *subscription) MarkEventAsSent(eventID string, events []types.ParsedEvent) {
	s.mutex.Lock()
	defer s.mutex.Unlock()

	eventsCopy := make([]types.ParsedEvent, len(events))
	copy(eventsCopy, events)
	s.sentEvents[eventID] = &eventsCopy
}

func (s *subscription) HasEventBeenSent(eventID string) bool {
	s.mutex.RLock()
	defer s.mutex.RUnlock()
	_, exists := s.sentEvents[eventID]
	return exists
}

func (s *subscription) AppendToSentEvent(eventID string, event types.ParsedEvent) {
	s.mutex.Lock()
	defer s.mutex.Unlock()

	if s.sentEvents[eventID] == nil {
		s.sentEvents[eventID] = &[]types.ParsedEvent{event}
	} else {
		*s.sentEvents[eventID] = append(*s.sentEvents[eventID], event)
	}
}

func (s *subscription) UpdateSentEvent(eventID string, event types.ParsedEvent) {
	s.mutex.Lock()
	defer s.mutex.Unlock()

	if s.sentEvents[eventID] != nil && len(*s.sentEvents[eventID]) > 0 {
		(*s.sentEvents[eventID])[0] = event
	}
}

func (s *subscription) Context() context.Context {
	return s.ctx
}

func (s *subscription) CreatedAt() time.Time {
	return s.createdAt
}

// FETCHED_EVENT batching methods
func (s *subscription) AddToFetchedBatch(events []types.ParsedEvent) {
	s.mutex.Lock()
	defer s.mutex.Unlock()
	
	s.fetchedBatch = append(s.fetchedBatch, events)
}

func (s *subscription) GetFetchedBatch() [][]types.ParsedEvent {
	s.mutex.RLock()
	defer s.mutex.RUnlock()
	
	// Return a copy to prevent external mutation
	result := make([][]types.ParsedEvent, len(s.fetchedBatch))
	for i, events := range s.fetchedBatch {
		eventsCopy := make([]types.ParsedEvent, len(events))
		copy(eventsCopy, events)
		result[i] = eventsCopy
	}
	return result
}

func (s *subscription) ClearFetchedBatch() {
	s.mutex.Lock()
	defer s.mutex.Unlock()
	
	s.fetchedBatch = s.fetchedBatch[:0] // Reset slice but keep capacity
}



func (s *subscription) IsInBatchingMode() bool {
	s.mutex.RLock()
	defer s.mutex.RUnlock()
	
	return s.batchingMode
}

func (s *subscription) SetBatchingMode(batching bool) {
	s.mutex.Lock()
	defer s.mutex.Unlock()
	
	s.batchingMode = batching
}

// subscriptionRegistry implements the SubscriptionRegistry interface
type subscriptionRegistry struct {
	subscriptions map[string]Subscription
	mutex         sync.RWMutex
}

// NewSubscriptionRegistry creates a new subscription registry
func NewSubscriptionRegistry() SubscriptionRegistry {
	return &subscriptionRegistry{
		subscriptions: make(map[string]Subscription),
	}
}

func (r *subscriptionRegistry) Create(id string) Subscription {
	r.mutex.Lock()
	defer r.mutex.Unlock()

	// Close existing subscription if it exists
	if existing, exists := r.subscriptions[id]; exists {
		existing.Cancel()
	}

	sub := NewSubscription(id)
	r.subscriptions[id] = sub
	return sub
}

func (r *subscriptionRegistry) Get(id string) (Subscription, bool) {
	r.mutex.RLock()
	defer r.mutex.RUnlock()

	sub, exists := r.subscriptions[id]
	return sub, exists
}

func (r *subscriptionRegistry) Remove(id string) {
	r.mutex.Lock()
	defer r.mutex.Unlock()

	if sub, exists := r.subscriptions[id]; exists {
		sub.Cancel()
		delete(r.subscriptions, id)
	}
}

func (r *subscriptionRegistry) Count() int {
	r.mutex.RLock()
	defer r.mutex.RUnlock()
	return len(r.subscriptions)
}

func (r *subscriptionRegistry) List() []string {
	r.mutex.RLock()
	defer r.mutex.RUnlock()

	ids := make([]string, 0, len(r.subscriptions))
	for id := range r.subscriptions {
		ids = append(ids, id)
	}
	return ids
}

func (r *subscriptionRegistry) Cleanup() {
	r.mutex.Lock()
	defer r.mutex.Unlock()

	for id, sub := range r.subscriptions {
		if sub.IsCancelled() {
			delete(r.subscriptions, id)
		}
	}
}

File: src/model/nostr-rust/src/network/subscriptions/subscription.rs
use crate::network::subscriptions::interfaces::SubscriptionTrait;
use crate::types::*;
use async_trait::async_trait;
use std::collections::HashMap;
use std::sync::Arc;
use tokio::sync::{oneshot, RwLock};

pub struct Subscription {
    id: String,
    sent_events: Arc<RwLock<HashMap<String, Vec<ParsedEvent>>>>,
    fetched_batch: Arc<RwLock<Vec<Vec<ParsedEvent>>>>,
    batching_mode: Arc<RwLock<bool>>,
    cancelled: Arc<RwLock<bool>>,
    cancel_tx: Arc<RwLock<Option<oneshot::Sender<()>>>>,
}

impl Subscription {
    pub fn new(id: String) -> Self {
        Self {
            id,
            sent_events: Arc::new(RwLock::new(HashMap::new())),
            fetched_batch: Arc::new(RwLock::new(Vec::new())),
            batching_mode: Arc::new(RwLock::new(false)),
            cancelled: Arc::new(RwLock::new(false)),
            cancel_tx: Arc::new(RwLock::new(None)),
        }
    }

    pub async fn with_cancel_channel(self, cancel_tx: oneshot::Sender<()>) -> Self {
        {
            let mut tx = self.cancel_tx.write().await;
            *tx = Some(cancel_tx);
        }
        self
    }
}

#[async_trait]
impl SubscriptionTrait for Subscription {
    fn id(&self) -> &str {
        &self.id
    }

    async fn is_cancelled(&self) -> bool {
        *self.cancelled.read().await
    }

    async fn cancel(&self) {
        let mut cancelled = self.cancelled.write().await;
        *cancelled = true;

        let mut tx_guard = self.cancel_tx.write().await;
        if let Some(tx) = tx_guard.take() {
            let _ = tx.send(());
        }
    }

    async fn get_sent_events(&self) -> HashMap<String, Vec<ParsedEvent>> {
        self.sent_events.read().await.clone()
    }

    async fn mark_event_as_sent(&self, event_id: &str, events: Vec<ParsedEvent>) {
        let mut sent_events = self.sent_events.write().await;
        sent_events.insert(event_id.to_string(), events);
    }

    async fn has_event_been_sent(&self, event_id: &str) -> bool {
        let sent_events = self.sent_events.read().await;
        sent_events.contains_key(event_id)
    }

    async fn add_to_fetched_batch(&self, events: Vec<ParsedEvent>) {
        let mut batch = self.fetched_batch.write().await;
        batch.push(events);
    }

    async fn get_fetched_batch(&self) -> Vec<Vec<ParsedEvent>> {
        self.fetched_batch.read().await.clone()
    }

    async fn clear_fetched_batch(&self) {
        let mut batch = self.fetched_batch.write().await;
        batch.clear();
    }

    async fn is_in_batching_mode(&self) -> bool {
        *self.batching_mode.read().await
    }

    async fn set_batching_mode(&self, batching: bool) {
        let mut mode = self.batching_mode.write().await;
        *mode = batching;
    }
}
